relay-server/src/metrics_extraction/transactions/mod.rs

```diff
@@ -375,6 +380,76 @@ fn extract_transaction_metrics_inner(
     Ok(())
 }
 
+fn extract_span_metrics(
+    aggregator_config: &AggregatorConfig,
+    event: &mut Event,
+    metrics: &mut Vec<Metric>, // output parameter
+) -> Result<(), ExtractMetricsError> {
+    // TODO(iker): measure the performance of this whole method
+
+    if event.ty.value() != Some(&EventType::Transaction) {
+        return Ok(());
+    }
+    let (Some(_), Some(&end)) = (event.start_timestamp.value(), event.timestamp.value()) else {
+        relay_log::debug!("failed to extract the start and the end timestamps from the event");
+        return Err(ExtractMetricsError::MissingTimestamp);
+    };
+
+    let Some(timestamp) = UnixTimestamp::from_datetime(end.into_inner()) else {
+        relay_log::debug!("event timestamp is not a valid unix timestamp");
+        return Err(ExtractMetricsError::InvalidTimestamp);
+    };
+
+    // Validate the transaction event against the metrics timestamp limits. If the metric is too
+    // old or too new, we cannot extract the metric and also need to drop the transaction event
+    // for consistency between metrics and events.
+    if !aggregator_config.timestamp_range().contains(&timestamp) {
+        relay_log::debug!("event timestamp is out of the valid range for metrics");
+        return Err(ExtractMetricsError::InvalidTimestamp);
+    }
+
+    if let Some(spans) = event.spans.value_mut() {
+        // Collect the shared tags for all the metrics and spans on this transaction
+        let mut shared_tags = BTreeMap::new();
+
+        if let Some(transaction_name) = event.transaction.value() {
+            shared_tags.insert("transaction".to_owned(), transaction_name.to_owned());
+        }
+        // TODO(iker): must add shared tags: environment, op, release
+
+        for annotated_span in spans {
+            if let Some(span) = annotated_span.value_mut() {
+                // TODO(iker): add span-specific tags here
+
+                if let Some(user) = event.user.value() {
+                    if let Some(value) = get_eventuser_tag(user) {
+                        metrics.push(Metric::new_mri(
+                            MetricNamespace::Spans,
+                            "user",
+                            MetricUnit::None,
+                            MetricValue::set_from_str(&value),
+                            timestamp,
+                            shared_tags.clone(),
+                        ));
+                    }
+                }
+
+                // Add to the span all the tags that exist in the metrics
+                let span_tags = span.tags.get_or_insert_with(BTreeMap::new);
+                for (key, value) in &shared_tags {
+                    // NOTE(iker): if the tag already existed before, we are overwriting it
+                    span_tags.insert(
+                        key.to_owned(),
+                        Annotated::new(JsonLenientString::from(value.to_owned())),
+                    );
+                }
```

Comment:
Instead of this you might be able to use [`extend`](https://doc.rust-lang.org/std/collections/struct.BTreeMap.html#method.extend)

something like 
```suggestion
                // Add to the span all the tags that exist in the metrics
                let mut span_tags = span.tags.get_or_insert_with(BTreeMap::new);
                span_tags.extend(shared_tags);
```
Did not test it , but I guess it should work.

Reply:
The type of the value of `BTreeMap` is `String`, and `span_tags`'s value is of type `Annotated<JsonLenientString>`, so afaik it's not possible to extend it.

Reply:
It's possib;e `map` the data and use that iterator to extend the `span_tags` in one go, which also will be must prettier and faster.

Reply:
I still think this looks better 😄 , which also could be done only once, before the metrics are generated and the tags then populated in all the future metric instances. 
```suggestion
                // Add to the span all the tags that exist in the metrics
                let span_tags = span.tags.get_or_insert_with(BTreeMap::new);
                span_tags.extend(
                    shared_tags
                        .clone()
                        .into_iter()
                        .map(|(k, v)| (k, Annotated::new(JsonLenientString::from(v)))),
                );

                
```

Reply:
Looks better, thank you! Done in [00fdeac](https://github.com/getsentry/relay/pull/2069/commits/00fdeacc95ca81455552756c3eed5f3e04771147).